imap_processing/hit/l0/decom_hit.py

```diff
@@ -0,0 +1,464 @@
+        # TODO log issue
+        print(f"Invalid seq_flgs found: {seq_flgs}")
+        return False
+
+    # Check that sequence counters are sequential
+    if not np.all(np.diff(src_seq_ctrs) == 1):
+        # TODO log issue
+        print(f"Non-sequential src_seq_ctr found: {src_seq_ctrs}")
+        return False
+
+    return True
+
+
+def update_ccsds_header_data(sci_dataset: xr.Dataset) -> xr.Dataset:
+    """
+    Update dimensions of CCSDS header fields.
+
+    The CCSDS header fields contain 1D arrays with
+    values from all the packets in the file. This
+    function updates the dimension for these fields
+    to use sc_tick instead of epoch. sc_tick is the
+    time the packet was created.
+
+    Parameters
+    ----------
+    sci_dataset : xr.Dataset
+        Xarray dataset containing HIT science packets
+        from a CCSDS file.
+
+    Returns
+    -------
+    sci_dataset : xr.Dataset
+        Updated xarray dataset.
+    """
+    # sc_tick contains spacecraft time per packet
+    sci_dataset.coords["sc_tick"] = sci_dataset["sc_tick"]
+    sci_dataset = sci_dataset.swap_dims({"epoch": "sc_tick"})
+    return sci_dataset
+
+
+def assemble_science_frames(sci_dataset: xr.Dataset) -> xr.Dataset:
+    """
+    Group packets into science frames.
+
+    HIT science frames (data from 1 minute) consist of 20 packets.
+    These are assembled from the binary science_data field in the
+    xarray dataset, which is a 1D array of science data from all
+    packets in the file, using packet sequence grouping flags.
+
+        First packet has a sequence flag = 1
+        Next 18 packets have a sequence flag = 0
+        Last packet has a sequence flag = 2
+
+    The science frame is further categorized into
+    L1A data products -> count rates and event data.
+
+        The first six packets contain count rates data
+        The last 14 packets contain pulse height event data
+
+    These groups are added to the dataset as count_rates_binary
+    and pha_binary.
+
+    Parameters
+    ----------
+    sci_dataset : xr.Dataset
+        Xarray Dataset for science data (APID 1252).
+
+    Returns
+    -------
+    sci_dataset : xr.Dataset
+        Updated xarray dataset with binary count rates and pulse
+        height event data per valid science frame added as new
+        data variables.
+    """
+    # TODO: Figure out how to handle partial science frames at the
+    #  beginning and end of CCSDS files. These science frames are split
+    #  across CCSDS files and still need to be processed. Only discard
+    #  incomplete science frames in the middle of the CCSDS file.
+    #  The code currently skips all incomplete science frames.
+
+    # Initialize lists to store data from valid science frames
+    count_rates_binary = []
+    pha_binary = []
+    epoch_science_frame = []
+
+    science_frame_start = 0
+    while science_frame_start + 20 <= len(sci_dataset.epoch):
```

Comment:
I feel like you may want to approach this differently. You may want to find indices of all 1s and 2 and then loop through those and check if those are valid science frame or not. If you start incrementing with 20 from beginning, you may miss good packets due to being off from where you should start. I am happy to jump on call if it helps.

Reply:
Yes, this would be helpful! I'll find a time to chat more about this

Reply:
I updated this to use numpy.where and numpy.diff. Please take a look when you get a chance!